libsolidity/codegen/ContractCompiler.cpp

```diff
@@ -1361,11 +1361,20 @@ void ContractCompiler::appendModifierOrFunctionCode()
 	m_context.setModifierDepth(m_modifierDepth);
 }
 
-void ContractCompiler::appendStackVariableInitialisation(VariableDeclaration const& _variable)
+void ContractCompiler::appendStackVariableInitialisation(
+	VariableDeclaration const& _variable,
+	bool _provideDefaultValue
+)
 {
 	CompilerContext::LocationSetter location(m_context, _variable);
 	m_context.addVariable(_variable);
-	CompilerUtils(m_context).pushZeroValue(*_variable.annotation().type);
+	if (!_provideDefaultValue && _variable.type()->dataStoredIn(DataLocation::Memory))
+	{
+		solAssert(_variable.type()->sizeOnStack() == 1, "");
+		m_context << u256(0);
+	}
+	else
+		CompilerUtils(m_context).pushZeroValue(*_variable.annotation().type);
```

Comment:
Wouldn't it be better to handle this in `pushZeroValue()`, specifically only in places where memory actually gets allocated?

Reply:
Not sure what you mean by "this piece of code", but `pushZeroValue` is a bit too low-level for me. It works as it currently is because there is no way to access a variable on the right-hand-side of its declaration. As soon as you can access it, you hav to provide a valid zero value.

Reply:
I meant the part that I selected for this comment, i.e. the whole newly added condition.

> It works as it currently is because there is no way to access a variable on the right-hand-side of its declaration.

I think that all `pushZeroValue()` would really need to check is whether the variable is stored in memory and if it's dynamically sized. I see that it already has access to this information via `referenceType->location()` and `arrayType->isDynamicallySized()`.

Oh, and I mean that `pushZeroValue()` would now have to accept the `_provideDefaultValue` argument too.

Anyway, this is just a minor suggestion so feel free to disregard it if you think it breaks the current design. It just looks to me like we're making an assumption about when `pushZeroValue()` allocates memory and when it doesn't. We could instead just tell it that the allocation can be skipped.

Reply:
I don't think there are so many situations where we can know that the value will be overwritten right away. Most of the time, we won't call `pushZeroValue` at all if we know it is not needed. In the present code, we need to push something on the stack, that's why we cannot just remove it.